dvc/repo/datasets.py

```diff
@@ -0,0 +1,350 @@
+        for name, (_, ds) in self._spec.items():
+            if ds["url"] == url:
+                return self.get(name)
+        return None
+
+    def __contains__(self, name: object) -> bool:
+        return name in self._spec
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._spec)
+
+    def __len__(self) -> int:
+        return len(self._spec)
+
+    @cached_property
+    def _spec(self) -> dict[str, tuple[str, dict[str, Any]]]:
+        return {
+            dataset["name"]: (path, dataset)
+            for path, datasets in self.repo.index._datasets.items()
+            for dataset in datasets
+        }
+
+    @cached_property
+    def _lock(self) -> dict[str, Optional[dict[str, Any]]]:
+        datasets_lock = self.repo.index._datasets_lock
+
+        def find(path, name) -> Optional[dict[str, Any]]:
+            lock = datasets_lock.get(path, [])
+            return next((dataset for dataset in lock if dataset["name"] == name), None)
+
+        return {ds["name"]: find(path, name) for name, (path, ds) in self._spec.items()}
+
+    def _reset(self):
+        self.__dict__.pop("_spec", None)
+        self.__dict__.pop("_lock", None)
+
+    def _spec_from_info(self, info: dict[str, Any]) -> Spec:
+        klasses: dict[str, type[Spec]] = {
+            "dvc": DVCDataset,
+            "dvcx": DVCXDataset,
+            "url": URLDataset,
+        }
+        cls = klasses[info["type"]]
+        return cls.from_dict(info)
+
+    def _lock_from_info(self, info: Optional[dict[str, Any]] = None) -> Optional[Lock]:
+        klasses: dict[str, type[Lock]] = {
+            "dvc": DVCDatasetLock,
+            "dvcx": DVCXDatasetLock,
+            "url": URLDatasetLock,
+        }
+
+        info = info or {}
+        if cls := klasses.get(info.get("type", None)):
+            return cls.from_dict(info)
+        return None
+
+    def add(
+        self, url: str, name: str, manifest_path: str = "dvc.yaml", **kwargs
+    ) -> Dataset:
+        kwargs.update({"url": url, "name": name})
+        spec = self._spec_from_info(kwargs)
+        dataset = Dataset(manifest_path=manifest_path, spec=spec)
+        dataset = dataset.update(self.repo)
+
+        self.dump(dataset)
+        return dataset
+
+    def update(self, name, **kwargs) -> tuple[Dataset, Dataset]:
+        dataset = self[name]
+        new = dataset.update(self.repo, **kwargs)
+
+        self.dump(new, old=dataset)
+        return dataset, new
+
+    def _dump_spec(self, manifest_path: str, spec: Spec) -> None:
+        spec_data = spec.to_dict()
+        project_file = ProjectFile(self.repo, manifest_path)
+        project_file.dump_dataset(spec_data)
+
+    def _dump_lock(self, manifest_path: str, lock: Lock) -> None:
+        lock_data = lock.to_dict()
+        lockfile = Lockfile(self.repo, Path(manifest_path).with_suffix(".lock"))
+        lockfile.dump_dataset(lock_data)
+
+    def dump(self, dataset: Dataset, old: Optional[Dataset] = None) -> None:
```

Comment:
This provides a _diff_ update to avoid accidentally modifying `dvc.yaml` or `dvc.lock`.